tools/building.py

```diff
@@ -1298,8 +1298,23 @@ def is_bitcode(filename):
 
 
 def is_wasm(filename):
-  magic = open(filename, 'rb').read(4)
-  return magic == b'\0asm'
+  if not os.path.exists(filename):
+    return False
+  header = open(filename, 'rb').read(webassembly.HEADER_SIZE)
+  return header == webassembly.MAGIC + webassembly.VERSION
+
+
+def is_wasm_dylib(filename):
+  """Detech wasm dynamic libraries by the presence of the "dylink" custom section."""
+  if not is_wasm(filename):
+    return False
+  module = webassembly.Module(filename)
```

Comment:
If we made `webassembly.Module` throw instead of asserting when the file isn't valid wasm, then we could skip the `if not is_wasm(filename)` above, and avoid opening the file twice.

Reply:
Interesting.. I'm not huge fan of exceptions but you are right,  this might be a place where it would make sense.  maybe as a followup?

Reply:
Followup is ok. I think exceptions are good sometimes (especially in languages like Python where using them in a few places doesn't cost much), and especially in constructors.